File: Python/Problems/Problems/problem539.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set
from collections import Counter

from .problem import Problem


class Problem539(Problem):
    def test(self) -> bool:
        timePoints = ["23:59", "00:00"]

        expected = 1

        result = self.findMinDifference(timePoints)

        print(result)

        return result == expected

    def findMinDifference(self, timePoints: List[str]) -> int:
        total = 24 * 60
        pres = [False] * total
        for t in timePoints:
            minutes = 600 * (ord(t[0]) - ord('0')) + 60 * (ord(t[1]) - ord('0')) + 10 * (ord(t[3]) - ord('0')) + (ord(t[4]) - ord('0'))
            if pres[minutes]:
                return 0
            pres[minutes] = True

        l = 0
        while not pres[l]:
            l += 1
        r = total - 1
        while not pres[r]:
            r -= 1
        result = l + total - r
        prev = l
        for i in range(l + 1, r + 1):
            if (pres[i]):
                result = min(result, i - prev)
                prev = i

        return result

File: C++/Problems/Problems/problem_539.cpp
#include "problem_539.h"

bool problem_539::test() {
    std::vector<std::string> timePoints { "23:59", "00:00" };

    int expected = 1;

    auto result = findMinDifference(timePoints);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_539::findMinDifference(std::vector<std::string> &timePoints) {
    int total = 24 * 60;
    std::vector pres(total, false);
    for (auto t : timePoints)
    {
        int minutes = 600 * (t[0] - '0') + 60 * (t[1] - '0') + 10 * (t[3] - '0') + (t[4] - '0');
        if (pres[minutes])
            return 0;
        pres[minutes] = true;
    }
    int l = 0;
    while (!pres[l])
        l++;
    int r = total - 1;
    while (!pres[r])
        r--;
    int result = l + total - r;
    int prev = l;
    for (int i = l + 1; i <= r; ++i)
    {
        if (pres[i])
        {
            result = std::min(result, i - prev);
            prev = i;
        }
    }

    return result;
}
